### PR TITLE
Add another method to either use user name or email in extracting the user

### PR DESCRIPTION
**Pull request by fanyui**
## Description of what I changed

I in addition of getiting a user only with email for password reseting have added another method that uses either email or username as in sql where username="username" or email = "email"

## Checklist: I completed these to help reviewers :)

- [ x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [ x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by fanyui** [comment]
Hello @wluyima I have used the approached you told me to use last time by creating a branch and each time making a pull to the passwordreset with this updated pull. My pull request is not passing all checks i've tried figuring out why but i have not been able to find out. I also had to squash all the commit but still it didn't pass the check.

File: NOTICE.md [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+<tr class="b">
+<td>javax.servlet</td>
+<td><a class="externalLink" href="http://servlet-spec.java.net">javax.servlet-api</a></td>
+<td>3.0.1</td>
+<td>jar</td>
+<td><a class="externalLink" href="https://glassfish.dev.java.net/nonav/public/CDDL+GPL.html">CDDL + GPLv2 with classpath exception</a></td></tr></table></div></div>
+      </div>
+    </div>
+  </body>
+</html>
```
Comment (wluyima) [inline NOTICE.md]: Why is this here? Can you please make sure your branch is up to date with openmrs/passwordreset and not openmrs/master?

File: api/src/main/java/org/openmrs/api/UserService.java [hunk 2]
```diff
@@ -102,6 +102,25 @@
...
+	
+	/**
+	 * gets the user with the specified username or email address
+	 * 
+	 * @param email Or username to look up user with
+	 * @return requested user
+	 * @throws APIException
+	 */
+	@Authorized( { PrivilegeConstants.GET_USERS })
+	public User getUserByEmailOrUsername(String emailOrName) throws APIException;
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/UserService.java]: I don't think you need this since you already have `getUserByUsername` and you've added `getUserByEmail`.

File: api/src/main/java/org/openmrs/validator/UserValidator.java [hunk 3]
```diff
@@ -193,4 +196,17 @@ public boolean isUserNameAsEmailValid(String username) {
...
+	/**
+	 * @return true if email is valid or null
+	 * @param email
+	 */
+	public boolean isEmailValid(String email) {
+		if (StringUtils.isBlank(email)) {
+			return true;
+		}
+		
+		Matcher matcher = EMAIL_PATTERN.matcher(email);
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/validator/UserValidator.java]: You can combine these 2 lines into 1

File: api/src/main/java/org/openmrs/validator/UserValidator.java [hunk 4]
```diff
@@ -193,4 +196,17 @@ public boolean isUserNameAsEmailValid(String username) {
...
 		return matcher.matches();
 	}
+	
+	/**
+	 * @return true if email is valid or null
+	 * @param email
+	 */
+	public boolean isEmailValid(String email) {
+		if (StringUtils.isBlank(email)) {
+			return true;
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/validator/UserValidator.java]: If it's blank then why do you return true?

**Comment by wluyima** [comment]
Where are the unit tests? It's best practice to always include unit tests for any change you make.

File: api/src/main/java/org/openmrs/api/UserService.java [hunk 5]
```diff
@@ -102,6 +102,25 @@
 	@Authorized( { PrivilegeConstants.GET_USERS })
 	public User getUserByUsername(String username) throws APIException;
 	
+	
+	/**
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/UserService.java]: Missing javadocs

File: api/src/main/java/org/openmrs/api/UserService.java [hunk 6]
```diff
@@ -102,6 +102,25 @@
...
+	 * @param email 
+	 * @return requested user
+	 * @throws APIException
+	 */
+	@Authorized( { PrivilegeConstants.GET_USERS })
+	public User getUserByEmail(String email) throws APIException;
+	
+	/**
+	 * gets the user with the specified username or email address
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/UserService.java]: Start with upper case latter please

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateUserDAO.java [hunk 7]
```diff
@@ -106,6 +106,16 @@ public User getUserByUsername(String username) {
...
 	}
 	
+	
+	/**
+	 * @see org.openmrs.api.UserService#getUserByEmail(java.lang.String)
+	 */
+	@Override
+	@SuppressWarnings("unchecked")
+	public User getUserByEmail(String email) {
+		return (User) sessionFactory.getCurrentSession().createCriteria(User.class).add(Restrictions.eq("email", email)).uniqueResult();
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateUserDAO.java]: I'd think this should be case insensitive

File: api/src/main/java/org/openmrs/validator/UserValidator.java [hunk 8]
```diff
@@ -126,6 +126,9 @@ public void validate(Object obj, Errors errors) {
 					errors.rejectValue("username", "error.username.pattern");
 				}
 			}
+			if(!isEmailValid(user.getEmail())) {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/validator/UserValidator.java]: I think email is optional so you should only validate the email when it's provided

File: api/src/main/resources/liquibase-schema-only.xml [hunk 9]
```diff
@@ -1828,6 +1828,9 @@
             </column>
             <column name="username" type="varchar(50)"/>
             <column name="password" type="varchar(128)"/>
+            <column name="email" type="varchar(255)">
+            	<constraints nullable="true" unique="true"/>
```
Comment (wluyima) [inline api/src/main/resources/liquibase-schema-only.xml]: I believe true is the default value for nullable, so this is redundant

File: api/src/main/java/org/openmrs/api/UserService.java (hunk 2 above)
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/UserService.java]: I was thinking of a single query to hit the db with an or conditioning since i found out from the validation file that email were accepted as username and instead of calling two different methods when i either an email or password is supplied for reseting password i just call this generic method to get the user then i now use the email associated with the user.

File: api/src/main/java/org/openmrs/validator/UserValidator.java (hunk 4 above)
Comment (fanyui) [reply api/src/main/java/org/openmrs/validator/UserValidator.java]: Null is accepted so preventing the validation from failing i let it know it should allow null but if something is supplied then validate with email pattern.

File: api/src/main/java/org/openmrs/validator/UserValidator.java (hunk 8 above)
Comment (fanyui) [reply api/src/main/java/org/openmrs/validator/UserValidator.java]: Ok that's true

File: api/src/main/java/org/openmrs/validator/UserValidator.java (hunk 4 above)
Comment (wluyima) [reply api/src/main/java/org/openmrs/validator/UserValidator.java]: This is a utility method an I guess which should return null if email is blank, the code that calls this should first check if not blank before calling it

File: api/src/main/java/org/openmrs/validator/UserValidator.java [hunk 10]
```diff
@@ -193,4 +196,17 @@ public boolean isUserNameAsEmailValid(String username) {
 		Matcher matcher = EMAIL_PATTERN.matcher(username);
 		return matcher.matches();
 	}
+	
+	/**
+	 * @return true if email is valid or null
+	 * @param email
+	 */
+	public boolean isEmailValid(String email) {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/validator/UserValidator.java]: This should probably be private if you don't call it outside of this class

File: api/src/main/java/org/openmrs/api/UserService.java (hunk 2 above)
Comment (wluyima) [reply api/src/main/java/org/openmrs/api/UserService.java]: Which validation file?

Comment (fanyui) [reply api/src/main/java/org/openmrs/api/UserService.java]: here is the file i saw. https://github.com/openmrs/openmrs-core/blob/master/api/src/main/java/org/openmrs/validator/UserValidator.java#L179

**Comment by fanyui** [comment]
@wluyima  I have updated the pull request with this commit https://github.com/openmrs/openmrs-core/pull/2679/commits/108b0f72c6a5de5dc96e7ffc4cbc0caa3cb8553a

File: api/src/main/java/org/openmrs/api/UserService.java (hunk 2 above)
Comment (wluyima) [reply api/src/main/java/org/openmrs/api/UserService.java]: Username and email are different, I don't think I understand what you mean

File: api/src/main/java/org/openmrs/validator/UserValidator.java [hunk 11]
```diff
@@ -193,4 +199,13 @@ public boolean isUserNameAsEmailValid(String username) {
 		Matcher matcher = EMAIL_PATTERN.matcher(username);
 		return matcher.matches();
 	}
+	
+	/**
+	 * @return true if email is valid or null
+	 * @param email
+	 */
+	private boolean isEmailValid(String email) {
+		Matcher matcher = EMAIL_PATTERN.matcher(email);
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/validator/UserValidator.java]: This can be combined into a single line

File: api/src/main/resources/messages.properties [hunk 12]
```diff
@@ -88,6 +88,7 @@ error.patientProgram.enrolledDateDateCannotBeInFuture=Enrolled date can't be in
 error.patientProgram.completionDateCannotBeInFuture=Completed date can't be in future, Please select a date on or before today
 error.context.null=contextDAO is null
 error.rank.notPositiveInteger=Rank can only be a positive integer
+error.email.invalid = "Invalid Email address"
```
Comment (wluyima) [inline api/src/main/resources/messages.properties]: You don't need the quotes

File: api/src/main/java/org/openmrs/User.java [hunk 13]
```diff
@@ -53,6 +53,8 @@
 	
 	private String email;
 	
+	private String userActivationKey;
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/User.java]: rename this to just `activationKey`

File: api/src/main/java/org/openmrs/User.java [hunk 14]
```diff
@@ -425,6 +427,20 @@ public String getEmail() {
 	public void setEmail(String email) {
 		this.email = email;
 	}
+
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/User.java]: Add @since to getters and setters, the version would be 2.3 possibly

File: api/src/main/java/org/openmrs/validator/UserValidator.java [hunk 15]
```diff
@@ -131,6 +133,13 @@ public void validate(Object obj, Errors errors) {
 					errors.rejectValue("email", "error.email.invalid");
 				}
 			}
+			
+			if (!StringUtils.isBlank(user.getUserActivationKey())) {
+				if(!isUserActivationKeyValid(user.getUserActivationKey())) {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/validator/UserValidator.java]: the activation key is system generated, why do you have to validate it against a certain pattern?

**Comment by wluyima** [comment]
Please can you address all the initial comments I added to the initial commit?

And can you always make sure you include unit tests for all code changes you make?

File: api/src/main/java/org/openmrs/validator/UserValidator.java [hunk 16]
```diff
@@ -41,6 +41,8 @@
 	private static final Pattern EMAIL_PATTERN = Pattern
 	        .compile("^.+@.+\\..+$");
 	
+	private static final Pattern USER_ACTIVATION_KEY = Pattern.compile("^\\d+(.[A-Za-z0-9]+)$");
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/validator/UserValidator.java]: I don't think this is required, I thought the key should also include the ttl

**Comment by fanyui** [comment]
Yes i  the ttl is included in the activation key ie i concatenated it with the activation key reason why i had to validate making sure that the time in secondes come first before the hash of the token both combined to form the activation key. Or am i to add another field seperate to the database for the ttl?.

**Comment by wluyima** [comment]
I guess you don't have to validate it since you're the one generating it anyways, plus that pattern might not work if you include the ttl in the token

**Comment by fanyui** [comment]
Ok the pattern did work with ttl but i will remove the validation completely. And i had address the innitial comment already before this last commit

File: api/src/main/resources/liquibase-schema-only.xml [hunk 17]
```diff
@@ -1828,6 +1828,12 @@
             </column>
             <column name="username" type="varchar(50)"/>
             <column name="password" type="varchar(128)"/>
+            <column name="email" type="varchar(255)">
+            	<constraints unique="true"/>
+            </column>
+            <column name="user_activation_key" type="varchar(50)">
+            	<constraints unique="true"/>
```
Comment (wluyima) [inline api/src/main/resources/liquibase-schema-only.xml]: I don't think you need a unique key constraint on this column

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateUserDAO.java [hunk 18]
```diff
@@ -122,7 +123,7 @@ public User getUserByEmail(String email) {
 	@Override
 	@SuppressWarnings("unchecked")
 	public User getUserByActivationKey(String activationKey) {
-		return (User) sessionFactory.getCurrentSession().createCriteria(User.class).add(Restrictions.eq("user_activation_key", activationKey)).uniqueResult();	
+		return (User) sessionFactory.getCurrentSession().createCriteria(User.class).add(Restrictions.ilike("activation_key", activationKey, MatchMode.START)).uniqueResult();
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateUserDAO.java]: I think for this to be more correct you need to actually compare the keys after you have stripped off the ttl

**Comment by wluyima** [comment]
You are still missing some unit tests, can you please add them?

**Comment by fanyui** [comment]
Hello @wluyima  please can you take a look again i just updated it pull request with a commit.

File: api/src/main/java/org/openmrs/api/UserService.java [hunk 19]
```diff
@@ -102,6 +102,25 @@
...
+	 * gets user by email
+	 * @param email 
+	 * @return requested user
+	 * @throws APIException
+	 */
+	@Authorized( { PrivilegeConstants.GET_USERS })
+	public User getUserByEmail(String email) throws APIException;
+	
+	/**
+	 * gets user using user activation key
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/UserService.java]: Use uppercase for the first letter of the javadocs

File: api/src/main/java/org/openmrs/api/db/UserDAO.java [hunk 20]
```diff
@@ -40,6 +40,16 @@
 	 */
 	public User getUserByUsername(String username) throws DAOException;
 	
+	/**
+	 * @see org.openmrs.api.UserService#getUserByEmail(java.lang.String)
+	 */
+	public User getUserByEmail(String email) throws DAOException;
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/UserDAO.java]: I know we add throws clause for DAOException historically but it's not necessary since it's a runtime exception

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateUserDAO.java [hunk 21]
```diff
@@ -106,6 +107,25 @@ public User getUserByUsername(String username) {
...
+		return (User) sessionFactory.getCurrentSession().createCriteria(User.class).add(Restrictions.eq("email", email).ignoreCase()).uniqueResult();	
+	}
+	
+	/**
+	 * @see org.openmrs.api.UserService#getUserByActivationKey(java.lang.String)
+	 */
+	@Override
+	@SuppressWarnings("unchecked")
+	public User getUserByActivationKey(String activationKey) {
+		return (User) sessionFactory.getCurrentSession().createCriteria(User.class).add(Restrictions.like("activationKey", activationKey, MatchMode.START)).uniqueResult();
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateUserDAO.java]: To make this concrete you need to compare the keys for equality before returning the match

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 22]
```diff
@@ -695,5 +695,29 @@ public String getSecretQuestion(User user) throws APIException {
...
+	}
 	
+	/**
+	 * @see org.openmrs.api.UserService#getUserByActivationKey(java.lang.String)
+	 * 
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public User getUserByActivationKey(String activationKey) throws APIException {
+
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Can you remove this space, did you format your code?

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 23]
```diff
@@ -695,5 +695,29 @@ public String getSecretQuestion(User user) throws APIException {
...
+	
+	/**
+	 * @see org.openmrs.api.UserService#getUserByEmail(java.lang.String)
+	 * 
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public User getUserByEmail(String email) throws APIException {
+
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Can you remove this space?

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 24]
```diff
@@ -1351,4 +1351,32 @@ public void changePasswordUsingSecretAnswer_shouldNotUpdatePasswordIfSecretIsNot
...
+
+		assertNotNull("User with email hank.williams@gmail not found in database", user);
+	}
+	/*
+	 * @see UserService#getUserByActivationKey(String)
+	 */
+	@Test
+	public void getUserByActivationKey_shouldGetUserUsingActivationKey() {
+		executeDataSet(XML_FILENAME);
+
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java]: Can you remove this space please?

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 25]
```diff
@@ -1351,4 +1351,32 @@ public void changePasswordUsingSecretAnswer_shouldNotUpdatePasswordIfSecretIsNot
...
+	 * @see UserService#getUserByActivationKey(String)
+	 */
+	@Test
+	public void getUserByActivationKey_shouldGetUserUsingActivationKey() {
+		executeDataSet(XML_FILENAME);
+		
+		User user = userService.getUserByActivationKey("XbklsieiskeuztecEnPf");
+		
+		assertNotNull("User with activationKey ", user); 
+		if(user!=null) {
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java]: This is bad practice in a unit test, you should be asserting that it's not null instead

Comment (wluyima) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java]: This is redundant because on the line before you asserted that's not null

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 26]
```diff
@@ -1351,4 +1351,32 @@ public void changePasswordUsingSecretAnswer_shouldNotUpdatePasswordIfSecretIsNot
...
+		assertNotNull("User with email hank.williams@gmail not found in database", user);
+	}
+	/*
+	 * @see UserService#getUserByActivationKey(String)
+	 */
+	@Test
+	public void getUserByActivationKey_shouldGetUserUsingActivationKey() {
+		executeDataSet(XML_FILENAME);
+		
+		User user = userService.getUserByActivationKey("XbklsieiskeuztecEnPf");
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java]: Can you store the key in a variable since you reuse it later

File: api/src/test/java/org/openmrs/api/db/UserDAOTest.java [hunk 27]
```diff
@@ -57,6 +57,8 @@ public void runBeforeEachTest() {
 		userJoe.addName(name);
 		userJoe.setUsername("juser");
 		userJoe.setDateCreated(new Date());
+		userJoe.setActivationKey("XbklsieiskeuztecEnPf:1529341342");
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/db/UserDAOTest.java]: What is the point of these 2 changes?

File: api/src/test/java/org/openmrs/api/db/UserDAOTest.java [hunk 28]
```diff
@@ -178,4 +180,16 @@ public void isSecretAnswer_shouldReturnFalseWhenTheAnswerDoesNotMatch() {
 		
 	}
 	
+	@Test
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/db/UserDAOTest.java]: These are duplicate tests to those in UserServiceTest

**Comment by fanyui** [comment]
i've done some updates again with respect to the comments

File: api/src/main/java/org/openmrs/api/UserService.java [hunk 29]
```diff
@@ -102,6 +102,25 @@
 	@Authorized( { PrivilegeConstants.GET_USERS })
 	public User getUserByUsername(String username) throws APIException;
 	
+	
+	/**
+	 * Gets user by email
+	 * @param email
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/UserService.java]: Missing javadocs for the param

File: api/src/main/java/org/openmrs/api/UserService.java [hunk 30]
```diff
@@ -102,6 +102,25 @@
...
+	 * @param email 
+	 * @return requested user
+	 * @throws APIException
+	 */
+	@Authorized( { PrivilegeConstants.GET_USERS })
+	public User getUserByEmail(String email) throws APIException;
+	
+	/**
+	 * Gets user using user activation key
+	 * @param activation key
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/UserService.java]: Missing javadocs for the param

File: api/src/main/java/org/openmrs/api/UserService.java [hunk 31]
```diff
@@ -102,6 +102,25 @@
...
 	
+	
+	/**
+	 * Gets user by email
+	 * @param email 
+	 * @return requested user
+	 * @throws APIException
+	 */
+	@Authorized( { PrivilegeConstants.GET_USERS })
+	public User getUserByEmail(String email) throws APIException;
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/UserService.java]: Get rid of the throws APIException since it's a runtime exception

File: api/src/main/java/org/openmrs/api/UserService.java [hunk 32]
```diff
@@ -102,6 +102,25 @@
...
+	
+	/**
+	 * Gets user using user activation key
+	 * @param activation key 
+	 * @return requested user
+	 * @throws APIException
+	 */
+	@Authorized( { PrivilegeConstants.GET_USERS })
+	public User getUserByActivationKey(String activationKey) throws APIException;
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/UserService.java]: Same here, you can remove the throws clause

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateUserDAO.java [hunk 33]
```diff
@@ -106,6 +107,34 @@ public User getUserByUsername(String username) {
...
+	 * @see org.openmrs.api.UserService#getUserByActivationKey(java.lang.String)
+	 */
+	@Override
+	@SuppressWarnings("unchecked")
+	public User getUserByActivationKey(String activationKey) {
+		User ret = null;
+		User user = (User) sessionFactory.getCurrentSession().createCriteria(User.class).add(Restrictions.like("activationKey", activationKey, MatchMode.START)).uniqueResult();	
+		if(user != null) {
+			String[] tokens = user.getActivationKey().split(":");
+			if(tokens[0].compareTo(activationKey) == 0) {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateUserDAO.java]: Why not use equals? Please use equals

File: api/src/main/java/org/openmrs/validator/UserValidator.java (hunk 10 above)
Comment (wluyima) [reply api/src/main/java/org/openmrs/validator/UserValidator.java]: Did you see this comment? Please check through for all unaddressed comments

Comment (fanyui) [reply api/src/main/java/org/openmrs/validator/UserValidator.java]: I used this method in UserValidatorTest So i needed it to be public

Comment (wluyima) [reply api/src/main/java/org/openmrs/validator/UserValidator.java]: That's not the right way to do it, you should be testing the validator that method is internal logic to the validator, what happens inside the validator should be treated like a blackbox. Am saying you should test that if a user has an invalid email, the validator fails and vice versa rather than exposing isEmailValid

**Comment by fanyui** [comment]
I had modified with respect to the comments you made. please do look at it again. Also i had to add dependency for EmailValidator.

File: api/pom.xml [hunk 34]
```diff
@@ -235,6 +235,11 @@
 		   <groupId>net.sf.ehcache</groupId>
 		   <artifactId>ehcache</artifactId>
        </dependency>
+       <dependency>
+		    <groupId>commons-validator</groupId>
+		    <artifactId>commons-validator</artifactId>
+		    <version>1.4.0</version>
```
Comment (wluyima) [inline api/pom.xml]: I thought this was already in core, sorry I didn't realize it, probably wasn't worth the effort.

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateUserDAO.java [hunk 35]
```diff
@@ -106,6 +107,34 @@ public User getUserByUsername(String username) {
...
+		return (User) sessionFactory.getCurrentSession().createCriteria(User.class).add(Restrictions.eq("email", email).ignoreCase()).uniqueResult();	
+	}
+	
+	/**
+	 * @see org.openmrs.api.UserService#getUserByActivationKey(java.lang.String)
+	 */
+	@Override
+	@SuppressWarnings("unchecked")
+	public User getUserByActivationKey(String activationKey) {
+		User ret = null;
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateUserDAO.java]: Line 126 is redundant

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateUserDAO.java [hunk 36]
```diff
@@ -106,6 +107,34 @@ public User getUserByUsername(String username) {
...
+		User user = (User) sessionFactory.getCurrentSession().createCriteria(User.class).add(Restrictions.like("activationKey", activationKey, MatchMode.START)).uniqueResult();	
+		if(user != null) {
+			String[] tokens = user.getActivationKey().split(":");
+			if(tokens[0].equals(activationKey)) {
+				ret = user;
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateUserDAO.java]: It's cleaner to replace this with **return user;** remove line 126 and replace line134 with **return null;**

File: api/src/main/java/org/openmrs/validator/UserValidator.java [hunk 37]
```diff
@@ -186,11 +190,19 @@ public boolean isUserNameValid(String username) {
...
-		Matcher matcher = EMAIL_PATTERN.matcher(username);
-		return matcher.matches();
 	}
+	
+	/**
+	 * @return true if email is valid or false otherwise
+	 * @param email
+	 */
+	private boolean isEmailValid(String email) {
+		EmailValidator emailValidator = EmailValidator.getInstance();
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/validator/UserValidator.java]: Can you combine these 2 lines into 1

**Comment by wluyima** [comment]
Almost there! I added a few more comments, should be quick fixes and get it merged

File: api/src/main/java/org/openmrs/validator/UserValidator.java [hunk 38]
```diff
@@ -186,11 +190,15 @@ public boolean isUserNameValid(String username) {
 	 * @should return true if email valid
 	 */
 	public boolean isUserNameAsEmailValid(String username) {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/validator/UserValidator.java]: Where do you call this method? You should not have an method as public, validator logic should be internal to a validator

**Comment by fanyui** [comment]
I didn't  create  the method I only converted it to use email validate
since it was manually being validated.

**Comment by wluyima** [comment]
Ok.

When commenting can you please add the responses inline in the PR next to my comments rather than here
